sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java

```diff
@@ -0,0 +1,284 @@
+                    FixedWindows.of(Duration.standardSeconds(WINDOW_DURATION))));
+    inputCollection.apply(
+        ParDo.of(
+            new DoFn<KV<String, String>, Void>() {
+              @ProcessElement
+              public void processElement(ProcessContext c, BoundedWindow window) {
+                LOGGER.debug(
+                    "*** ELEMENT: (%s,%s) *** with timestamp %s in window %s",
+                    c.element().getKey(),
+                    c.element().getValue(),
+                    c.timestamp().toString(),
+                    window.toString());
+              }
+            }));
+
+    // elements have the same key and collection is divided into windows,
+    // so Count.perKey values are the number of elements in windows
+    PCollection<KV<String, Long>> countInput =
+        inputCollection.apply(
+            "Count elements in windows before applying batchingParDo",
+            Count.<String, String>perKey());
+    PAssert.that("Wrong number of elements in windows before BatchingParDo", countInput)
+        .satisfies(new CheckValuesFn(WINDOW_DURATION));
+
+    PCollection<KV<String, String>> outputCollection =
+        inputCollection
+            .apply(batchingParDo)
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+
+    PAssert.that("All elements have not been processed", outputCollection)
+        .satisfies(new CheckAllElementsProcessingFn());
+
+    // elements have the same key and collection is divided into windows,
+    // so Count.perKey values are the number of elements in windows
+    PCollection<KV<String, Long>> countOutput =
+        outputCollection.apply(
+            "Count elements in windows after applying batchingParDo",
+            Count.<String, String>perKey());
+
+    PAssert.that("Wrong number of elements in windows after BatchingParDo", countOutput)
+        .satisfies(new CheckValuesFn(WINDOW_DURATION));
+    pipeline.run().waitUntilFinish();
+  }
+
+  private boolean checkAllElementsProcessing(
+      Iterable<KV<String, String>> listToCheck, Processing processing) {
+    for (KV<String, String> element : listToCheck) {
+      if (processing == Processing.PROCESSED) {
+        if (!element.getValue().matches(".*2")) {
+          return false;
+        }
+      } else {
+        if (element.getValue().matches(".*2")) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private enum Processing {
+    PROCESSED,
+    UNPROCESSED
+  }
+
+  private class CheckAllElementsProcessingFn
+      implements SerializableFunction<Iterable<KV<String, String>>, Void> {
+    @Override
+    public Void apply(Iterable<KV<String, String>> input) {
+      assertTrue(
+          "all elements of the collection have not been processed ",
+          checkAllElementsProcessing(input, Processing.PROCESSED));
+      return null;
+    }
+  }
+
+  private class CheckValuesFn implements SerializableFunction<Iterable<KV<String, Long>>, Void> {
```

Comment:
Can this be inlined into the appropriate test? You use it twice in that test, with the same configuration both times.